File: dlna-dmr/src/main/java/com/android/cast/dlna/dmr/service/AudioRenderController.java
package com.android.cast.dlna.dmr.service;

import android.content.Context;
import android.media.AudioManager;

import org.fourthline.cling.model.types.UnsignedIntegerFourBytes;
import org.fourthline.cling.model.types.UnsignedIntegerTwoBytes;

/**
 *
 */
public final class AudioRenderController implements IRendererInterface.IAudioControl {

    private static final UnsignedIntegerTwoBytes VOLUME_MUTE = new UnsignedIntegerTwoBytes(0);
    private final UnsignedIntegerFourBytes mInstanceId;
    private final AudioManager mAudioManager;
    private UnsignedIntegerTwoBytes mLastVolume;
    private UnsignedIntegerTwoBytes mCurrentVolume;

    public AudioRenderController(Context context) {
        this(context, new UnsignedIntegerFourBytes(0));
    }

    public AudioRenderController(Context context, UnsignedIntegerFourBytes instanceId) {
        mInstanceId = instanceId;
        mAudioManager = (AudioManager) context.getSystemService(Context.AUDIO_SERVICE);
        final int MAX_MUSIC = mAudioManager.getStreamMaxVolume(AudioManager.STREAM_MUSIC);
        final int volume = mAudioManager.getStreamVolume(AudioManager.STREAM_MUSIC);
        mCurrentVolume = new UnsignedIntegerTwoBytes(volume * 100 / MAX_MUSIC);
        mLastVolume = new UnsignedIntegerTwoBytes(volume * 100 / MAX_MUSIC);
    }

    public UnsignedIntegerFourBytes getInstanceId() {
        return mInstanceId;
    }

    @Override
    public void setMute(String channelName, boolean desiredMute) {
        if (desiredMute) {
            mLastVolume = mCurrentVolume;
        }
        setVolume(channelName, desiredMute ? VOLUME_MUTE : mLastVolume);
    }

    @Override
    public boolean getMute(String channelName) {
        return getVolume(channelName).getValue() == 0L;
    }

    @Override
    public void setVolume(String channelName, UnsignedIntegerTwoBytes desiredVolume) {
        mCurrentVolume = desiredVolume;
        int volume = desiredVolume.getValue().intValue();
        int adjustVolume = volume * mAudioManager.getStreamMaxVolume(AudioManager.STREAM_MUSIC) / 100;
        mAudioManager.setStreamVolume(AudioManager.STREAM_MUSIC, adjustVolume, AudioManager.FLAG_PLAY_SOUND | AudioManager.FLAG_SHOW_UI);
    }

    @Override
    public UnsignedIntegerTwoBytes getVolume(String channelName) {
        return mCurrentVolume;
    }
}

File: dlna-dmr/src/main/java/com/android/cast/dlna/dmr/service/AudioRenderController.kt
package com.android.cast.dlna.dmr.service

import android.content.Context
import android.media.AudioManager
import com.android.cast.dlna.core.Logger
import org.fourthline.cling.model.types.UnsignedIntegerTwoBytes

/**
 *
 */
class AudioRenderController constructor(context: Context) : AudioControl {

    private val muteVolume = UnsignedIntegerTwoBytes(0)
    private val audioManager: AudioManager
    private var currentVolume: UnsignedIntegerTwoBytes

    init {
        audioManager = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
        val maxMusicVolume = audioManager.getStreamMaxVolume(AudioManager.STREAM_MUSIC)
        val volume = audioManager.getStreamVolume(AudioManager.STREAM_MUSIC)
        currentVolume = UnsignedIntegerTwoBytes(volume * 100L / maxMusicVolume)
    }

    override val logger: Logger = Logger.create("AudioRenderController")

    override fun setMute(channelName: String, desiredMute: Boolean) {
        super.setMute(channelName, desiredMute)
        setVolume(channelName, if (desiredMute) muteVolume else currentVolume)
    }

    override fun getMute(channelName: String): Boolean = getVolume(channelName).value == 0L

    override fun setVolume(channelName: String, desiredVolume: UnsignedIntegerTwoBytes) {
        super.setVolume(channelName, desiredVolume)
        currentVolume = desiredVolume
        val volume = desiredVolume.value.toInt()
        val adjustVolume = volume * audioManager.getStreamMaxVolume(AudioManager.STREAM_MUSIC) / 100
        audioManager.setStreamVolume(AudioManager.STREAM_MUSIC, adjustVolume, AudioManager.FLAG_PLAY_SOUND or AudioManager.FLAG_SHOW_UI)
    }

    override fun getVolume(channelName: String): UnsignedIntegerTwoBytes =
        UnsignedIntegerTwoBytes(audioManager.getStreamVolume(AudioManager.STREAM_MUSIC) * 100L / audioManager.getStreamMaxVolume(AudioManager.STREAM_MUSIC))
}
